### PR TITLE
Support recursive NamedTuple classes

### LINKED ISSUE #9397: Support recursive NamedTuple
## 🐛 Bug Report

NamedTuple's appear to behave in unexpected ways with recursion, especially when compared to frozen dataclasses (which are rather similar, as far as the user is concerned).

## To Reproduce

```python
from __future__ import annotations
from typing import NamedTuple, Optional

class A(NamedTuple):
  a: Optional[A]
```

Produces a "mypy-sample1.py:5: error: Cannot resolve name "A" (possible cyclic definition)". However,

```python
from __future__ import annotations
from typing import Optional
from dataclasses import dataclass

@dataclass(frozen=True)
class A:
  a: Optional[A]
```

Passes MyPy just fine, and as far as a user is concerned, presents a very similar surface for MyPy to type check (i.e. both have a read-only attribute a which is either None or of class A).

## Expected Behavior

The expected behaviour would be that both NamedTuple and frozen dataclass behave the same, as NamedTuple and frozen dataclasses present logically similar behaviour with regards to the attribute 'a'.

## Actual Behavior

In addition to the odd behaviour with cycle checker triggering on NamedTuple but not on dataclasses, it's possible to fool the cycle checker completely and have MyPy crash by recursing too far (e.g. issue #8695 and although that issue has been reported with a recursive type that could never be instantiated, I've encountered the same behaviour with type definitions that can be instantiated, but not in a minimal example friendly format). I think this specific behaviour is the root cause, as the real problem is that for some reason NamedTuple's are behaving differently to what would be expected.

## Your Environment

- Mypy version used: mypy 0.782
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.8.2
- Operating system and version: Linux

### PR DESCRIPTION
**Pull request by niklasl**
### Description

Recursive named tuple classes like this now pass:
```python
from __future__ import annotations
from typing import NamedTuple

class Node(NamedTuple):
    parent: Node | None
```
(Instead of causing a `Cannot resolve name "MyNamedTuple" (possible cyclic definition)` error.)

This is done by letting `mypy.semanal_namedtuple.analyze_namedtuple_classdef` complete recursive type references in a subsequent pass.

Fixes #9397.

## Test Plan

Tests are updated to reflect this new support. (I did not add my own use case above since the existing tests already cover this feature which previously caused the error (along with many more special cases).)

## Questions

I believe this change is good, but it raises some questions.

1. The now conditional call to `add_symbol`  in `mypy.semanal.prepare_class_def` does add to the somewhat complex situation already in there, regarding func scope nesting. I tried to make the logic clear, but would you prefer me to attempt to improve it further?

   (I did note the already present `TODO` in there. It seems hard to get at the clarity without untangling that, which I think is a separate task. I initially added a flag to avoid the double check, which I can't say made it more understandable, but rather indicated the already noted need for a refactoring.)

3. Related but distinct, an odd error cropped up in `check-incremental.test::testIncrementalWithDifferentKindsOfNestedTypesWithinMethod` which made `lookup_fully_qualified` throw an error:

       E   AssertionError: Cannot find component 'NT1@11' for 'b.NT1@11'

   Somehow this is prevented by a call to `reveal_type` (regardless of value). Did I stumble upon a separate issue, or is there an obvious mistake in this change?

4. In `check-namedtuple.test::testSelfRefNT4` it is shown that `reveal_type` for an item-accessed named tuple value now returns `builtins.object` instead of `Any`. While it is beyond the scope of this PR to fix type-inference for named tuple item access in general, would you consider the resulting semantics a regression or an improvement?

5. I used a `for ... else` in `analyze_namedtuple_classdef` since (IMHO) it's a good fit and made things clearer. But if you're adverse to that language feature I can change it (and if so reasonably also advice against it in the coding conventions).

### CODE REVIEW

File: mypy/semanal_namedtuple.py [hunk 1]
```diff
@@ -392,28 +398,30 @@ def build_namedtuple_typeinfo(self,
...
-        info.tuple_type = tuple_base
         info.line = line
+
+        return info
+
+    def _complete_namedtuple_typeinfo(self,
+                                  info: TypeInfo,
+                                  items: List[str],
+                                  types: List[Type],
+                                  default_items: Mapping[str, Expression]):
```
Comment (AlexWaygood) [inline mypy/semanal_namedtuple.py:421]:
```suggestion
                                  default_items: Mapping[str, Expression]) -> None:
```

**Comment by AlexWaygood** [comment]
I'm not a mypy core dev, and know very little about the parts of mypy that provide namedtuple support, so I'm not a good reviewer for this PR, sorry :)

**Review by hauntsaninja — COMMENTED** [review]
Thank you for the PR, but superseded by https://github.com/python/mypy/pull/13371/files

**Comment by niklasl** [comment]
@hauntsaninja Thanks for the feedback! I'm glad progress is being made on this in any case (and at least I got some insight into the codebase from this attempt).

**Pull request closed without merging**